[PATCH 0/5] clean up meson configuration

takaswie opened the pull request:
This patchset cleans up current meson configuration.

```
Takashi Sakamoto (5):
  meson: obsolete unused variable for documentation
  meson: obsolete usage of deprecated meson.build_root()
  meson: evaluate private headers
  meson: simplify .pc generation for pkgconfig
  meson: suppress scan to marshaller codes for gir

 meson.build       |  2 +-
 src/meson.build   | 17 ++++++++---------
 tests/meson.build |  6 ++----
 3 files changed, 11 insertions(+), 14 deletions(-)
```

eli-schwartz commented on src/meson.build:
```diff
@@ -71,15 +71,12 @@ install_headers(headers,
 pkg = import('pkgconfig')
 pkg.generate(
-  version: meson.project_version(),
-  libraries: myself,
```
Hmm, I would suspect you still want to pass the primary library as the first posarg to pkg.generate?

takaswie replied:
Thanks for your comment.

Indeed. I overlooked that the first positional argument is available since v0.46. I think it's preferable since it's possible to parse ELF binary for libhinoko library to generate .pc file without passing explicit parameters in the latest version of meson:

https://mesonbuild.com/Pkgconfig-module.html

```
Since 0.46 a StaticLibrary or SharedLibrary object can optionally be passed as first positional argument. If one is provided a default value will be provided for all required fields of the pc file:

install_dir is set to pkgconfig folder in the same location than the provided library.
description is set to the project's name followed by the library's name.
name is set to the library's name.

Implicit dependencies
The exact rules followed to find dependencies that are implicitly added into the pkg-config file have evolved over time. Here are the rules as of Meson 0.49.0, previous versions might have slightly different behaviour.
...
```

I still wonder the reason that .pc file is generated expectedly without the argument, but later I'll force-push respined patchset which includes your indication.

eli-schwartz replied:
Is it generated expectedly?

I would expect it to be generated successfully, but to be missing the libhinoko library since it's not listed anywhere in the invocation.

takaswie replied:
> Is it generated expectedly?
>
> I would expect it to be generated successfully, but to be missing the libhinoko library since it's not listed anywhere in the invocation.

Indeed. I did take less care of differences between before/after the simplification. I can see explicit difference there...

```
$ diff -u hinoko.pc.prev hinoko.pc.new
--- hinoko.pc.prev      2022-03-10 12:02:43.902469857 +0900
+++ hinoko.pc.new       2022-03-10 12:05:15.969281391 +0900
@@ -5,5 +5,6 @@
 Name: hinoko
 Description: A library to operate OHCI 1394 controller for isochronous stream
 Version: 0.6.0
-Requires: glib-2.0, gobject-2.0
+Requires: glib-2.0, gobject-2.0 >= 2.34.0
+Libs: -L${libdir} -lhinoko
 Cflags: -I${includedir}/hinoko
```

Just now I force-pushed and move the HEAD of PR branch into d9489cbdf9cc . I'd like you to test that it can generate as we expected.

Thank you.

eli-schwartz replied:
In the latest force push of this PR, the generated pkg-config file is identical before and after this PR. So that looks good.

Looking at the change itself it seems like the correct use of `pkg.generate`. :)

Incidentally, you probably don't need to explicitly list glib-2.0 since you don't explicitly link to it and gobject-2.0 already has it as a public `Requires:`

You can pass `requires: dependencies` directly, you don't need to use the string names of them, which allows you to sync the names with the dependencies that you actually look up.

Ordinarily they would be listed by default anyway even if not referenced at all, because libhinoko links to them, but... that only adds them to `Requires.private` while you do need them to be public requires.

takaswie replied:
> In the latest force push of this PR, the generated pkg-config file is identical before and after this PR. So that looks good.
> 
> Looking at the change itself it seems like the correct use of `pkg.generate`. :)

It's nice.

> Incidentally, you probably don't need to explicitly list glib-2.0 since you don't explicitly link to it and gobject-2.0 already has it as a public `Requires:`
> 
> You can pass `requires: dependencies` directly, you don't need to use the string names of them, which allows you to sync the names with the dependencies that you actually look up.
> 
> Ordinarily they would be listed by default anyway even if not referenced at all, because libhinoko links to them, but... that only adds them to `Requires.private` while you do need them to be public requires.

To be honestly, I'm not so good at pkg-config itself as well as pkg-config module of meson. After a short research, the difference between `Requires` and `Requires.private` fields relates to `overlinking`, which brings useless rebuilt of application even in case of dynamic linking.

In case of `libhinoko`, libhinoko applications should not be rebuilt as long as built for dynamic linking when glib-2.0/gobject-2.0 is changed. I guess that the latest tools for linking supports function to prevent from overlinking issue, while for safe I conclude not to use `Require` field but to use `Require.private` field. In the meaning, it's convenient to use default behaviour of pkg-config module.

takaswie replied:
> In the latest force push of this PR, the generated pkg-config file is identical before and after this PR. So that looks good.
> 
> Looking at the change itself it seems like the correct use of `pkg.generate`. :)

It's nice.

> Incidentally, you probably don't need to explicitly list glib-2.0 since you don't explicitly link to it and gobject-2.0 already has it as a public `Requires:`
> 
> You can pass `requires: dependencies` directly, you don't need to use the string names of them, which allows you to sync the names with the dependencies that you actually look up.
> 
> Ordinarily they would be listed by default anyway even if not referenced at all, because libhinoko links to them, but... that only adds them to `Requires.private` while you do need them to be public requires.

To be honestly, I'm not so good at pkg-config itself as well as pkg-config module of meson. After a short research, the difference between `Requires` and `Requires.private` fields relates to `overlinking`, which brings useless rebuilt of application in case of dynamic linking.

In case of `libhinoko`, libhinoko applications should not be rebuilt as long as built for dynamic linking when glib-2.0/gobject-2.0 get changes (actually the applications need to use glib-2.0/gobject-2.0 API to operate stuffs in  libhinoko.). I guess that the latest tools for linking supports function to prevent from overlinking issue, while for safe I conclude not to use `Require` field but to use `Require.private` field. In the meaning, it's convenient to use default behaviour of pkg-config module.

takaswie replied:
I moved HEAD of PR branch into f20944e6822b from d9489cbdf9cc as merge candicate.

takaswie replied:
@eli-schwartz 

Thanks for your helpful comment. If you have enough time to care of the other libraries I maintain in alsa-project, your comment to them makes me really happy. They support g-i as well as libhinoko:

 * https://github.com/alsa-project/libhinawa
 * https://github.com/alsa-project/alsa-gobject/

Regards

takaswie committed:
meson: simplify .pc generation for pkgconfig

In meson build v0.46 or later, pkgconfig module automatically estimates
version number. Let's use it.

Additionally Eli Schwartz reports that usage of Require field brings
overlinking issue. The explicit parameter for require is deleted and I
rely on default behaviour of pkgconfig module.

Signed-off-by: Takashi Sakamoto <o-takashi@sakamocchi.jp>

takaswie committed:
meson: evaluate private headers

It's better to aware of private headers for compilation.

Signed-off-by: Takashi Sakamoto <o-takashi@sakamocchi.jp>

takaswie committed:
meson: obsolete usage of deprecated meson.build_root()

Meson build v0.56 deprecates meson.build_root(). This commit adds
alternative variable for the path to built stuffs.

Signed-off-by: Takashi Sakamoto <o-takashi@sakamocchi.jp>

takaswie committed:
meson: obsolete unused variable for documentation

I forgot to remove it when migrating to gi-docgen.

Signed-off-by: Takashi Sakamoto <o-takashi@sakamocchi.jp>

takaswie committed:
meson: suppress scan to marshaller codes for gir

The functions defined in marshaller are not available for applications,
therefore no need to scan them for gir.

Signed-off-by: Takashi Sakamoto <o-takashi@sakamocchi.jp>

takaswie commented:
Merged.
